intermediate_source/torch_export_tutorial.py

```diff
@@ -0,0 +1,533 @@
+#   operands and they must both return a single tensor with the same metadata (for example, ``dtype``, ``shape``, etc.).
+# - Branch functions cannot mutate input or global variables.
+# - Branch functions cannot access closure variables, except for ``self`` if the function is
+#   defined in the scope of a method.
+
+######################################################################
+# ..
+#     [NOTE] map is not documented at the moment
+#     We can also use ``map``, which applies a function across the first dimension
+#     of the first tensor argument.
+#
+#     from functorch.experimental.control_flow import map
+#
+#     def map_example(xs):
+#         def map_fn(x, const):
+#             def true_fn(x):
+#                 return x + const
+#             def false_fn(x):
+#                 return x - const
+#             return control_flow.cond(x.sum() > 0, true_fn, false_fn, [x])
+#         return control_flow.map(map_fn, xs, torch.tensor([2.0]))
+#
+#     exported_map_example= export(map_example, (torch.randn(4, 3),))
+#     inp = torch.cat((torch.ones(2, 3), -torch.ones(2, 3)))
+#     print(exported_map_example(inp))
+
+######################################################################
+# Constraints
+# -----------
+#
+# Ops can have different specializations/behaviors for different tensor shapes, so by default,
+# ``torch.export`` requires inputs to ``ExportedProgram`` to have the same shape as the respective
+# example inputs given to the initial ``torch.export`` call.
+# If we try to run the first ``ExportedProgram`` example with a tensor
+# with a different shape, we get an error:
+
+try:
+    exported_mod(torch.randn(10, 100), torch.randn(10, 100))
+except Exception:
+    tb.print_exc()
+
+######################################################################
+# We can modify the ``torch.export`` call to
+# relax some of these constraints. We use ``torch.export.dynamic_dim`` to
+# express shape constraints manually.
+#
+# ..
+#     [TODO] link to doc of dynamic_dim when it is available
+#
+# Using ``dynamic_dim`` on a tensor's dimension marks it as dynamic (i.e. unconstrained), and
+# we can provide additional upper and lower bound shape constraints.
+# The first argument of ``dynamic_dim`` is the tensor variable we wish
+# to specify a dimension constraint for. The second argument specifies
+# the dimension of the first argument the constraint applies to.
+# In the example below, our input
+# ``inp1`` has an unconstrained first dimension, but the size of the second
+# dimension must be in the interval (3, 18].
+
+from torch.export import dynamic_dim
+
+inp1 = torch.randn(10, 10)
+
+def constraints_example1(x):
+    x = x[:, 2:]
+    return torch.relu(x)
+
+constraints1 = [
+    dynamic_dim(inp1, 0),
+    3 < dynamic_dim(inp1, 1),
+    dynamic_dim(inp1, 1) <= 18,
+]
+
+exported_constraints_example1 = export(constraints_example1, (inp1,), constraints=constraints1)
```

Comment:
Might be helpful to also print(exported_constraints_example1.range_constraints) to show that the ranges for the symbolic shapes are there

Reply:
I cover printing out `range_constraints` and `equality_constraints` a little bit below.